libs/providers/ofrep-web/src/lib/ofrep-web-provider.ts

```diff
@@ -0,0 +1,246 @@
+    try {
+      this._context = newContext;
+      await this._evaluateFlags(newContext);
+    } catch (error) {
+      if (
+        error instanceof OpenFeatureError ||
+        error instanceof OFREPApiFetchError ||
+        error instanceof OFREPApiTooManyRequestsError ||
+        error instanceof OFREPApiUnauthorizedError ||
+        error instanceof OFREPForbiddenError
+      ) {
+        this.events?.emit(ClientProviderEvents.Error, { message: `${error.name}: ${error.message}` });
+        return;
+      }
+      this.events?.emit(ClientProviderEvents.Error, { message: `Unknown error: ${error}` });
+    }
+  }
+
+  /**
+   * onClose is called when the provider is closed, it will stop the polling if it is enabled.
+   */
+  onClose?(): Promise<void> {
+    this.stopPolling();
+    return Promise.resolve();
+  }
+
+  /**
+   * _evaluateFlags is a function that will call the bulk evaluate flags endpoint to get the flags values.
+   * @param context - the context to use for the evaluation
+   * @private
+   * @returns EvaluationStatus if the evaluation the API returned a 304, 200.
+   * @throws TargetingKeyMissingError if the API returned a 400 with the error code TargetingKeyMissing
+   * @throws InvalidContextError if the API returned a 400 with the error code InvalidContext
+   * @throws ParseError if the API returned a 400 with the error code ParseError
+   * @throws GeneralError if the API returned a 400 with an unknown error code
+   */
+  private async _evaluateFlags(context?: EvaluationContext | undefined): Promise<EvaluationStatus> {
+    const evalReq: EvaluationRequest = {
+      context,
+    };
+    const options: RequestOptions = {
+      headers: new Headers({
+        'Content-Type': 'application/json',
+      }),
+      ...(this._etag ? { headers: { 'If-None-Match': this._etag } } : {}),
+    };
+
+    const response = await this._ofrepAPI.postBulkEvaluateFlags(evalReq, options);
+    if (response.httpStatus === 304) {
+      // nothing has changed since last time, we are doing nothing
+      return EvaluationStatus.SUCCESS_NO_CHANGES;
+    }
+
+    if (response.httpStatus === 400) {
+      switch (response.value.errorCode) {
+        case EvaluationFailureErrorCode.TargetingKeyMissing:
+          throw new TargetingKeyMissingError();
+        case EvaluationFailureErrorCode.InvalidContext:
+          throw new InvalidContextError();
+        case EvaluationFailureErrorCode.ParseError:
+          throw new ParseError();
+        default:
+          throw new GeneralError();
+      }
+    }
+
+    if (response.httpStatus === 200) {
+      const bulkSuccessResp = response.value;
+      const newCache: { [key: string]: ResolutionDetails<FlagValue> } = {};
+
+      bulkSuccessResp.flags?.forEach((evalResp: EvaluationResponse) => {
+        if (evalResp.key === undefined) {
+          return;
+        }
+        const key = evalResp.key;
+        const evalSuccessResp = evalResp as EvaluationSuccessResponse;
```

Comment:
This cast should be replaced by handling error errors on here correctly.
Maybe we can have a cache of success and error responses or so?
When evaluating a flag with an error response we might want to return the default value and the error from the response or so.

Reply:
Good catch, I've reworked a bit that part and stored in the cache the evaluation in error during the bulk evaluation, to be able to throw the correct error.